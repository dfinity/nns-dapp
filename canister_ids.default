{
  "__Candid_UI": {
    "testnet": "qsgjb-riaaa-aaaaa-aaaga-cai"
  },
  "nns-dapp": {
    "testnet": "qhbym-qaaaa-aaaaa-aaafq-cai"
  }
}