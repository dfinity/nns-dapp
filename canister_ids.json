{
  "__Candid_UI": {
    "testnet": "sp3hj-caaaa-aaaaa-aaajq-cai"
  },
  "nns_ui": {
    "mainnet": "qoctq-giaaa-aaaaa-aaaea-cai",
    "messaging": "6qpux-vqaaa-aaaaa-qaalq-cai",
    "testnet": "qhbym-qaaaa-aaaaa-aaafq-cai"
  }
}
