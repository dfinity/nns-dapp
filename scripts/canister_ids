#!/usr/bin/env bash
set -euo pipefail

SOURCE_DIR="$(dirname "$(realpath "${BASH_SOURCE[0]}")")"

print_help() {
  cat <<-EOF

	Manage canister_ids.json.

	Usage: $(basename "$0") --export --network <network> (outputs to stdout)
	       $(basename "$0") --import                     (reads from stdin)
	       $(basename "$0") --import-from-index-html <URL> --network <network>
	       $(basename "$0") --remove --network <network> --canister <canister>
	EOF
}

# Source the clap.bash file ---------------------------------------------------
source "$SOURCE_DIR/clap.bash"
# Define options
clap.define short=e long=export desc="Export canister IDs" variable=EXPORT nargs=0 default="false"
clap.define short=i long=import desc="Import canister IDs from stdin" variable=IMPORT nargs=0 default="false"
clap.define long=import-from-index-html desc="URL of nns-dapp canister to import from" variable=IMPORT_FROM
clap.define short=r long=remove desc="Remove canister ID" variable=REMOVE nargs=0 default="false"
clap.define short=n long=network desc="dfx network to use" variable=NETWORK default="staging"
clap.define short=c long=canister desc="canister to remove" variable=CANISTER default="nns-dapp"
clap.define short=j long=json-file desc="Path to the canister_ids.json file" variable=JSON_FILE default="./canister_ids.json"
clap.define long=create desc="Create the json file if missing" variable=CREATE nargs=0
# Source the output file ----------------------------------------------------------
source "$(clap.build)"

if ! [ -f "${JSON_FILE:-}" ]; then
  if [ "${CREATE:-}" ]; then
    echo "{}" >"$JSON_FILE"
  else
    echo "${JSON_FILE:-} is missing.  Consider using: --create" >&2
    exit 1
  fi
fi

useCommand() {
  declare -g command
  if [ "${command:-}" ]; then
    echo "Only specify one of --import, --import-from-index-html, --export, --remove" >&2
    exit 1
  fi
  command="$1"
}

if [ "$EXPORT" = "true" ]; then
  useCommand "export"
fi

if [ "$IMPORT" = "true" ]; then
  useCommand "import"
fi

if [ "${IMPORT_FROM:-}" ]; then
  useCommand "import-from-index-html"
fi

if [ "$REMOVE" = "true" ]; then
  useCommand "remove"
fi

if ! [ "${command:-}" ]; then
  print_help
  exit 0
fi

if [ "$command" = "export" ]; then
  # This whole block can be done in a single line of jq script but it woudn't
  # be readable.
  canisters=$(jq -r 'keys|join(" ")' "$JSON_FILE")
  output="{}"
  for canister in $canisters; do
    id=$(jq -r --arg network "$NETWORK" --arg canister "$canister" '.[$canister][$network]?' "$JSON_FILE")
    if [ "$id" != "null" ]; then
      new_output=$(echo "$output" | jq --arg network "$NETWORK" --arg canister "$canister" --arg id "$id" '.[$canister][$network]=$id')
      output="$new_output"
    fi
  done
  echo "$output"
  exit
fi

if [ "$command" = "import" ]; then
  echo "Reading from stdin..."
  input="$(cat)"
  output="$(jq --argjson input "$input" '. * $input' "$JSON_FILE")"
  echo "$output" >"$JSON_FILE"
  exit
fi

if [ "$command" = "remove" ]; then
  # ShellCheck misfires, thinking $CANISTER isn't set.
  # shellcheck disable=SC2153
  old_canister_id="$(jq -r --arg network "$NETWORK" --arg canister "$CANISTER" '.[$canister][$network]' "$JSON_FILE")"

  if ! [ "$old_canister_id" ]; then
    echo "$CANISTER already didn't have an ID for $NETWORK."
    exit
  fi

  new_name="$NETWORK-before-$(date +%Y%m%d-%H%M%S)"
  tmp_file="$(mktemp)"
  jq --arg network "$NETWORK" --arg canister "$CANISTER" --arg new_name "$new_name" '.[$canister][$new_name] = .[$canister][$network] | del(.[$canister][$network])' "$JSON_FILE" >"$tmp_file"
  mv "$tmp_file" "$JSON_FILE"

  echo "Renamed $CANISTER.$NETWORK to $new_name."
  exit
fi

get_value_from_html() {
  key="$1"
  file="$2"
  pattern="$key=\"([^\"]+)\""
  # E.g. find a line like `data-own-canister-id="lf43c-fyaaa-aaaaa-aacva-cai"`
  # and outputs the value between quotes.
  grep -oE "$pattern" "$file" | sed -E "s/$pattern/\\1/" | head -1 || true
}

canister_id_from_url() {
  echo "$1" | sed -E 's@https://([^.]*)\..*@\1@'
}

if [ "$command" = "import-from-index-html" ]; then
  tmp_index_html="$(mktemp)"
  trap 'rm -f "$tmp_index_html"' EXIT
  curl -sS "$IMPORT_FROM" >"$tmp_index_html"
  if [ "$(file --brief --mime-type "$tmp_index_html")" = "application/gzip" ]; then
    mv "$tmp_index_html" "${tmp_index_html}.gz"
    gunzip "${tmp_index_html}.gz"
  fi

  identityServiceUrl="$(get_value_from_html data-identity-service-url "$tmp_index_html")"
  internetIdentity="$(canister_id_from_url "$identityServiceUrl")"

  snsAggregatorUrl="$(get_value_from_html data-sns-aggregator-url "$tmp_index_html")"
  snsAggregator="$(canister_id_from_url "$snsAggregatorUrl")"

  nnsIndex="$(get_value_from_html data-index-canister-id "$tmp_index_html")"
  ckbtcIndex="$(get_value_from_html data-ckbtc-index-canister-id "$tmp_index_html")"
  ckbtcLedger="$(get_value_from_html data-ckbtc-ledger-canister-id "$tmp_index_html")"
  ckbtcMinter="$(get_value_from_html data-ckbtc-minter-canister-id "$tmp_index_html")"
  ckethIndex="$(get_value_from_html data-cketh-index-canister-id "$tmp_index_html")"
  ckethLedger="$(get_value_from_html data-cketh-ledger-canister-id "$tmp_index_html")"
  ckusdcIndex="$(get_value_from_html data-ckusdc-index-canister-id "$tmp_index_html")"
  ckusdcLedger="$(get_value_from_html data-ckusdc-ledger-canister-id "$tmp_index_html")"
  snsWasm="$(get_value_from_html data-wasm-canister-id "$tmp_index_html")"
  nnsDapp="$(get_value_from_html data-own-canister-id "$tmp_index_html")"

  tmp_file="$(mktemp)"
  jq '
    .["nns-index"][$network] = $nnsIndex |
    .ckbtc_index[$network] = $ckbtcIndex |
    .ckbtc_ledger[$network] = $ckbtcLedger |
    .ckbtc_minter[$network] = $ckbtcMinter |
    .cketh_index[$network] = $ckethIndex |
    .cketh_ledger[$network] = $ckethLedger |
    .ckusdc_index[$network] = $ckusdcIndex |
    .ckusdc_ledger[$network] = $ckusdcLedger |
    .["nns-sns-wasm"][$network] = $snsWasm |
    .internet_identity[$network] = $internetIdentity |
    .["nns-dapp"][$network] = $nnsDapp |
    .sns_aggregator[$network] = $snsAggregator' \
    --arg network "$NETWORK" \
    --arg nnsIndex "$nnsIndex" \
    --arg ckbtcIndex "$ckbtcIndex" \
    --arg ckbtcLedger "$ckbtcLedger" \
    --arg ckbtcMinter "$ckbtcMinter" \
    --arg ckethIndex "$ckethIndex" \
    --arg ckethLedger "$ckethLedger" \
    --arg ckusdcIndex "$ckusdcIndex" \
    --arg ckusdcLedger "$ckusdcLedger" \
    --arg snsWasm "$snsWasm" \
    --arg internetIdentity "$internetIdentity" \
    --arg nnsDapp "$nnsDapp" \
    --arg snsAggregator "$snsAggregator" \
    "$JSON_FILE" >"$tmp_file"
  mv "$tmp_file" "$JSON_FILE"
fi
