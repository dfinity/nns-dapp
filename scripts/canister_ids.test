#!/usr/bin/env bash
set -euo pipefail

SOURCE_DIR="$(dirname "$(realpath "${BASH_SOURCE[0]}")")"

test_json_file=$(mktemp)

reset_test_json_file() {
  cat >"$test_json_file" <<-EOF
	{
	  "__Candid_UI": {
	    "staging": "tmxop-wyaaa-aaaaa-aaapa-cai"
	  },
	  "internet_identity": {
	    "staging": "wflfh-4yaaa-aaaaa-aaata-cai"
	  },
	  "nns-dapp": {
	    "mainnet": "qoctq-giaaa-aaaaa-aaaea-cai",
	    "staging": "4yr67-tiaaa-aaaaa-aabka-cai"
	  },
	  "nns-governance": {
	    "mainnet": "rrkah-fqaaa-aaaaa-aaaaq-cai"
	  }
	}
	EOF
}

test_index_html=$(mktemp canister.raw.XXXXXX)

# data-cketh-index-canister-id is deliberately missing so that we can see that
# it results in an empty string instead of a crash.
cat >"$test_index_html" <<-EOF
<html>
  <head>
  <meta name="nns-dapp-vars"
        data-api-host="https://fubar.dfinity.network"
        data-ckbtc-index-canister-id="tqtu6-byaaa-aaaaa-aaana-cai"
        data-ckbtc-ledger-canister-id="st75y-vaaaa-aaaaa-aaalq-cai"
        data-ckbtc-minter-canister-id="t6rzw-2iaaa-aaaaa-aaama-cai"
        data-cketh-ledger-canister-id="omy6t-mmaaa-aaaaa-qabgq-cai"
        data-cycles-minting-canister-id="rkp4c-7iaaa-aaaaa-aaaca-cai"
        data-dfx-network="staging"
        data-fetch-root-key="true"
        data-governance-canister-id="rrkah-fqaaa-aaaaa-aaaaq-cai"
        data-host="https://fubar.dfinity.network"
        data-identity-service-url="https://wqmuk-5qaaa-aaaaa-aaaqq-cai.fubar.dfinity.network"
        data-index-canister-id="bkyz2-fmaaa-aaaaa-qaaaq-cai"
        data-ledger-canister-id="ryjl3-tyaaa-aaaaa-aaaba-cai"
        data-own-canister-id="lf43c-fyaaa-aaaaa-aacva-cai"
        data-robots="&lt;meta name=&quot;robots&quot; content=&quot;noindex, nofollow&quot; /&gt;"
        data-sns-aggregator-url="https://p5kov-xqaaa-aaaaa-aacmq-cai.fubar.dfinity.network"
        data-static-host="https://fubar.dfinity.network"
        data-tvl-canister-id=""
        data-wasm-canister-id="qaa6y-5yaaa-aaaaa-aaafa-cai"
        data-own-canister-id="lf43c-fyaaa-aaaaa-aacva-cai">
  </head>
</html>
EOF

clean_up() {
  rm "$test_json_file"
  rm "$test_index_html"
}

trap 'clean_up' EXIT

: Test export

reset_test_json_file

if ! diff <("$SOURCE_DIR/canister_ids" --export --network staging --json-file "$test_json_file") <(
  cat <<-EOF
{
  "__Candid_UI": {
    "staging": "tmxop-wyaaa-aaaaa-aaapa-cai"
  },
  "internet_identity": {
    "staging": "wflfh-4yaaa-aaaaa-aaata-cai"
  },
  "nns-dapp": {
    "staging": "4yr67-tiaaa-aaaaa-aabka-cai"
  }
}
EOF
); then
  echo "ERROR exporting: See output of diff above."
  exit 1
fi

: Test import

reset_test_json_file

"$SOURCE_DIR/canister_ids" --import --json-file "$test_json_file" <<-EOF
{
  "internet_identity": {
    "staging": "new-ii-id"
  },
  "nns-dapp": {
    "staging": "new-nns-dapp-id"
  }
}
EOF

if ! diff "$test_json_file" <(
  cat <<-EOF
{
  "__Candid_UI": {
    "staging": "tmxop-wyaaa-aaaaa-aaapa-cai"
  },
  "internet_identity": {
    "staging": "new-ii-id"
  },
  "nns-dapp": {
    "mainnet": "qoctq-giaaa-aaaaa-aaaea-cai",
    "staging": "new-nns-dapp-id"
  },
  "nns-governance": {
    "mainnet": "rrkah-fqaaa-aaaaa-aaaaq-cai"
  }
}
EOF
); then
  echo "ERROR importing: See output of diff above."
  exit 1
fi

: Test import-from-index-html

reset_test_json_file

"$SOURCE_DIR/canister_ids" --import-from-index-html "file://$(realpath "$test_index_html")" --json-file "$test_json_file"

if ! diff "$test_json_file" <(
  cat <<-EOF
{
  "__Candid_UI": {
    "staging": "tmxop-wyaaa-aaaaa-aaapa-cai"
  },
  "internet_identity": {
    "staging": "wqmuk-5qaaa-aaaaa-aaaqq-cai"
  },
  "nns-dapp": {
    "mainnet": "qoctq-giaaa-aaaaa-aaaea-cai",
    "staging": "lf43c-fyaaa-aaaaa-aacva-cai"
  },
  "nns-governance": {
    "mainnet": "rrkah-fqaaa-aaaaa-aaaaq-cai"
  },
  "nns-index": {
    "staging": "bkyz2-fmaaa-aaaaa-qaaaq-cai"
  },
  "ckbtc_index": {
    "staging": "tqtu6-byaaa-aaaaa-aaana-cai"
  },
  "ckbtc_ledger": {
    "staging": "st75y-vaaaa-aaaaa-aaalq-cai"
  },
  "ckbtc_minter": {
    "staging": "t6rzw-2iaaa-aaaaa-aaama-cai"
  },
  "cketh_index": {
    "staging": ""
  },
  "cketh_ledger": {
    "staging": "omy6t-mmaaa-aaaaa-qabgq-cai"
  },
  "nns-sns-wasm": {
    "staging": "qaa6y-5yaaa-aaaaa-aaafa-cai"
  },
  "sns_aggregator": {
    "staging": "p5kov-xqaaa-aaaaa-aacmq-cai"
  }
}
EOF
); then
  echo "ERROR importing from URL: See output of diff above."
  exit 1
fi

: Test remove

reset_test_json_file

"$SOURCE_DIR/canister_ids" --remove --canister nns-dapp --network mainnet --json-file "$test_json_file"

# The removed entry was backed up with a key depending on the date/time so
# remove it before comapring.
if ! diff <(grep -v '"mainnet-before-........-......"' "$test_json_file") <(
  cat <<-EOF
{
  "__Candid_UI": {
    "staging": "tmxop-wyaaa-aaaaa-aaapa-cai"
  },
  "internet_identity": {
    "staging": "wflfh-4yaaa-aaaaa-aaata-cai"
  },
  "nns-dapp": {
    "staging": "4yr67-tiaaa-aaaaa-aabka-cai",
  },
  "nns-governance": {
    "mainnet": "rrkah-fqaaa-aaaaa-aaaaq-cai"
  }
}
EOF
); then
  echo "ERROR removing: See output of diff above."
  exit 1
fi

echo PASSED
