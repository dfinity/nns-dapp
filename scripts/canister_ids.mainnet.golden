{
  "nns-index": {
    "mainnet": "qhbym-qaaaa-aaaaa-aaafq-cai"
  },
  "ckbtc_index": {
    "mainnet": "n5wcd-faaaa-aaaar-qaaea-cai"
  },
  "ckbtc_ledger": {
    "mainnet": "mxzaz-hqaaa-aaaar-qaada-cai"
  },
  "ckbtc_minter": {
    "mainnet": "mqygn-kiaaa-aaaar-qaadq-cai"
  },
  "cketh_index": {
    "mainnet": "s3zol-vqaaa-aaaar-qacpa-cai"
  },
  "cketh_ledger": {
    "mainnet": "ss2fx-dyaaa-aaaar-qacoq-cai"
  },
  "ckusdc_index": {
    "mainnet": "xrs4b-hiaaa-aaaar-qafoa-cai"
  },
  "ckusdc_ledger": {
    "mainnet": "xevnm-gaaaa-aaaar-qafnq-cai"
  },
  "nns-sns-wasm": {
    "mainnet": "qaa6y-5yaaa-aaaaa-aaafa-cai"
  },
  "internet_identity": {
    "mainnet": "identity"
  },
  "nns-dapp": {
    "mainnet": "qoctq-giaaa-aaaaa-aaaea-cai"
  },
  "sns_aggregator": {
    "mainnet": "3r4gx-wqaaa-aaaaq-aaaia-cai"
  }
}
